job-server-api/src/spark.jobserver/NamedRddSupport.scala

```diff
@@ -6,115 +6,191 @@ import java.util.concurrent.atomic.AtomicReference
+   * It facilitates easy sharing of RDDs amongst jobs sharing the same SparkContext.
+   * If two jobs simultaneously tries to create an RDD with the same name,
+   * only one will win and the other will retrieve the same one.
    *
-   * If an RDD is requested by thread B while thread A is generating the RDD, thread B will block up to
-   * the duration specified by @timeout. If thread A finishes generating the RDD within that time, then
-   * thread B will get a reference to the newly-created RDD. If thread A does not finish generating the
-   * RDD within that time, then thread B will throw a timeout exception.
-   *
-   * @param name the unique name of the RDD. The uniqueness is scoped to the current SparkContext.
-   * @param rddGen a 0-ary function which will generate the RDD if it doesn't already exist.
-   * @param forceComputation if true, forces the RDD to be computed by calling count().
-   * @param storageLevel the storage level to persist the RDD with. Default: StorageLevel.MEMORY_ONLY.
-   * @param timeout if the RddManager doesn't respond within this timeout, an error will be thrown.
-   * @tparam T the generic type of the RDD.
-   * @return the RDD with the given name.
-   * @throws java.util.concurrent.TimeoutException if the request to the RddManager times out.
-   * @throws java.lang.RuntimeException wrapping any error that occurs within the generator function.
+   * Note that to take advantage of NamedRddSupport, a job must mix this in and use the APIs here instead of
+   * the native RDD `cache()`, otherwise we will not know about the names.
    */
-  def getOrElseCreate[T](name: String,
-                         rddGen: => RDD[T],
-                         forceComputation: Boolean = true,
-                         storageLevel: StorageLevel = defaultStorageLevel)
-                        (implicit timeout: Timeout = defaultTimeout): RDD[T]
+  trait _NamedRdds {
 
-  /**
-   * Gets an RDD with the given name if it already exists and is cached by the RddManager.
-   * If the RDD does not exist, None is returned.
-   *
-   * Note that a previously-known RDD could 'disappear' if it hasn't been used for a while, because the
-   * SparkContext garbage-collects old cached RDDs.
-   *
-   * @param name the unique name of the RDD. The uniqueness is scoped to the current SparkContext.
-   * @param timeout if the RddManager doesn't respond within this timeout, an error will be thrown.
-   * @tparam T the generic type of the RDD.
-   * @return the RDD with the given name.
-   * @throws java.util.concurrent.TimeoutException if the request to the RddManager times out.
-   */
-  def get[T](name: String)(implicit timeout: Timeout = defaultTimeout): Option[RDD[T]]
+    // Default timeout is 60 seconds. Hopefully that is enough
+    // to let most RDD/DataFrame generator functions finish.
+    val defaultTimeout = akka.util.Timeout(Duration(60, java.util.concurrent.TimeUnit.SECONDS))
```

Comment:
done, but again, to make this work, I had to change the type of the timeout parameter to `Option[Timeout]`